project/preprocess_dockerfile.sh

```diff
@@ -0,0 +1,30 @@
+#!/bin/bash
+echo "#!/bin/bash"
+echo "cat << EOF"
+regex='^(if|elif|else|fi).*$'
+regex2='.*\\$'
+regex3='(.*)(\$|`)(.*)'
+while IFS='' read -r line
+do
+    if [[ $line =~ $regex ]]; then
+        echo "EOF"
+        echo "$line" 
+	echo "cat << EOF"
+    else
+	if [[ $line =~ $regex2 ]]; then
+		line="$line\\"
+	fi    
+	if [[ $line =~ $regex3 ]]; then
+        	templine="${BASH_REMATCH[1]}"
+		result="\\${BASH_REMATCH[2]}${BASH_REMATCH[3]}"
+		while [[ $templine =~ $regex3 ]]; do
+			templine="${BASH_REMATCH[1]}"
+			result="\\${BASH_REMATCH[2]}${BASH_REMATCH[3]}$result"
+		done
+		line="$templine$result"
+    	fi
+        echo "$line"
+    fi  
+done <$(dirname "$BASH_SOURCE")/../Dockerfile.in
+echo "EOF"
+
```

Comment:
@tianon we may need to pick the lesser of two evils. As @brahmaroutu points out there are probably two choices:
1 - do what he's done here
2 - use heredocs but then we have to escape all $, \\, and \` chars
I guess the 3rd choice is:
3 - add IF/ENDIF support to Dockerfiles but @tiborvass wasn't thrilled with that one.
Can you think of another option?  Which of these are you leaning towards?